geopandas/tools/clip.py

```diff
@@ -37,41 +50,117 @@ def _clip_points(gdf, poly):
         The returned GeoDataFrame is a subset of gdf that intersects
         with poly.
     """
-    return gdf.iloc[gdf.sindex.query(poly, predicate="intersects")]
+    clipping_candidates = gdf.iloc[gdf.sindex.query(poly, predicate="intersects")]
+    point_indices = clipping_candidates.geom_type == "Point"
+    # For performance reasons points don't need to be intersected with poly
+    points = clipping_candidates[point_indices]
 
+    non_points = clipping_candidates[~point_indices]
+    non_points_intersected = _intersect_non_point_geometries_with_poly(non_points, poly)
+    return pd.concat([points, non_points_intersected])
 
-def _clip_line_poly(gdf, poly):
-    """Clip line and polygon geometry to the polygon extent.
 
-    Clip an input line or polygon to the polygon extent of the poly
-    parameter. Parts of Lines or Polygons that intersect the poly geometry are
-    extracted with associated attributes and returned.
+def _validate_input_gdf(gdf):
+    if not isinstance(gdf, (GeoDataFrame, GeoSeries)):
+        raise TypeError(
+            "'gdf' should be GeoDataFrame or GeoSeries, got {}".format(type(gdf))
+        )
 
-    Parameters
-    ----------
-    gdf : GeoDataFrame, GeoSeries
-        Line or polygon geometry that is clipped to poly.
 
-    poly : (Multi)Polygon
-        Reference polygon for clipping.
+def _validate_input_mask(mask):
+    if not isinstance(mask, (GeoDataFrame, GeoSeries, Polygon, MultiPolygon)):
+        raise TypeError(
+            "'mask' should be GeoDataFrame, GeoSeries or"
+            "(Multi)Polygon, got {}".format(type(mask))
+        )
 
-    Returns
-    -------
-    GeoDataFrame
-        The returned GeoDataFrame is a clipped subset of gdf
-        that intersects with poly.
-    """
-    gdf_sub = gdf.iloc[gdf.sindex.query(poly, predicate="intersects")]
 
-    # Clip the data with the polygon
-    if isinstance(gdf_sub, GeoDataFrame):
-        clipped = gdf_sub.copy()
-        clipped[gdf.geometry.name] = gdf_sub.intersection(poly)
+def _validate_inputs(gdf, mask):
+    _validate_input_gdf(gdf)
+    _validate_input_mask(mask)
+    _warn_if_crs_mismatch(gdf, mask)
+
+
+def _warn_if_crs_mismatch(gdf, mask):
+    if isinstance(mask, (GeoDataFrame, GeoSeries)):
+        if not _check_crs(gdf, mask):
+            _crs_mismatch_warn(gdf, mask, stacklevel=3)
+
+
+def _contains_geometrycollection(gdf):
+    return (gdf.geom_type == "GeometryCollection").any()
+
+
+def _get_number_of_unique_geometry_types(gdf):
+    return sum(gdf.geom_type.isin(geom_type).any() for geom_type in _GEOMETRY_TYPES)
```

Comment:
```suggestion
    return gdf.geom_type.nunique()
```

Isn't this the same (but faster)?

Again, we can use this line directly (esp. if the `nunique` solution works) where it is needed.

Reply:
So in this case it would actually not be the same as

```python
_POINT_TYPES = ["Point", "MultiPoint"]
_LINE_TYPES = ["LineString", "MultiLineString", "LinearRing"]
_POLYGON_TYPES = ["Polygon", "MultiPolygon"]
_GEOMETRY_TYPES = [_POINT_TYPES, _LINE_TYPES, _POLYGON_TYPES]
```

e.g. if there are only geometries of type `"Point"` and `"MultiPoint"`, this would count as 1 unique geometry type. I also find this rather strange, but this was the way it was done in the original code, and I did not want to introduce any further semantic changes.

Reply:
Maybe, to avoid confusion, I could change the name to `_GEOMETRY_TYPE_GROUPS = [_POINT_TYPE_GROUP, ...]` and then call the function `_get_number_of_unique_geometry_type_groups()`?

Reply:
Ah, yeah, I see it now. Just get rid of the function I'd say and keep the rest.

Reply:
Done